helix-core/src/test/java/org/apache/helix/manager/zk/TestParticipantManager.java

```diff
@@ -0,0 +1,182 @@
+        "MasterSlave",
+        true); // do rebalance
+
+    final String instanceName = "localhost_12918";
+    final MockParticipantManager manager =
+        new MockParticipantManager(ZK_ADDR, clusterName, instanceName);
+
+    manager.syncStart();
+
+    final LiveInstance liveInstance = accessor.getProperty(keyBuilder.liveInstance(instanceName));
+    final long originalCreationTime = liveInstance.getStat().getCreationTime();
+    final String originalSessionId = manager.getSessionId();
+
+    // Verify current live instance.
+    Assert.assertNotNull(liveInstance);
+    Assert.assertEquals(liveInstance.getEphemeralOwner(), originalSessionId);
+
+    final CountDownLatch startCountdown = new CountDownLatch(1);
+    final CountDownLatch endCountdown = new CountDownLatch(1);
+    final Semaphore semaphore = new Semaphore(0);
+    manager.addPreConnectCallback(
+        new BlockingPreConnectCallback(instanceName, startCountdown, endCountdown, semaphore));
+
+    // Expire S0 and new session S1 will be created.
+    ZkTestHelper.asyncExpireSession(manager.getZkClient());
+
+    // Wait for onPreConnect to start
+    semaphore.acquire();
+
+    // New session S1 should not be equal to S0.
+    Assert.assertFalse(originalSessionId.equals(manager.getSessionId()));
+
+    // Live instance should be gone as original session S0 is expired.
+    Assert.assertNull(accessor.getProperty(keyBuilder.liveInstance(instanceName)));
+
+    final String lastSessionId = manager.getSessionId();
+
+    // Expire S1 when S1 is blocked in onPreConnect().
+    // New session S2 will be created.
+    ZkTestHelper.asyncExpireSession(manager.getZkClient());
+
+    TestHelper.verify(
+        () -> !(ZKUtil.toHexSessionId(manager.getZkClient().getSessionId()).equals(lastSessionId)),
+        3000L);
+
+    // New session S2 should not be equal to S1.
+    final String latestSessionId = ZKUtil.toHexSessionId(manager.getZkClient().getSessionId());
+    Assert.assertFalse(lastSessionId.equals(latestSessionId));
+
+    // Proceed S1 to create live instance, which will fail.
+    startCountdown.countDown();
+
+    // Wait until S1's handling new session is completed.
+    semaphore.acquire();
+
+    // Live instance should not be created because zk session is expired.
+    Assert.assertNull(accessor.getProperty(keyBuilder.liveInstance(instanceName)),
+        "Live instance should not be created because zk session is expired!");
+
+    // Proceed S2 to create live instance.
+    endCountdown.countDown();
+
+    TestHelper.verify(() -> {
+      // Newly created live instance should be created by the latest session S2
+      // and have a new creation time.
+      LiveInstance newLiveInstance =
+          accessor.getProperty(keyBuilder.liveInstance(instanceName));
+      return newLiveInstance != null
+          && newLiveInstance.getStat().getCreationTime() != originalCreationTime
+          && newLiveInstance.getEphemeralOwner().equals(latestSessionId);
+    }, 2000L);
```

Comment:
Nit: we seem to be being inconsistent with wait times - could we use TestHelper's default wait time (if there is a constant defined already)?

Reply:
There is a constant wait time (20 seconds). 20s is kind of too long for this test because if waiting is longer than 2s, the test should definitely fail. But anyway, I will adopt your suggestion to use the constant. Maybe we can add at least one more constant to TestHelper like `SHORT_WAIT_DURATION`